### ISSUE TITLE
[YSQL] ysql_dump fails due to high  memory usage

### ISSUE THREAD

**Issue by sushantrmishra** [yugabyte/yugabyte-db#14212]
Jira Link: [DB-3677](https://yugabyte.atlassian.net/browse/DB-3677)
### Description

Taking backup on  SQL schema with 70k tables fails  due to high memory usage consumption PostgreSQL backend causing OOM.

From memory context dumps reported,  caching of the catalog is highest memory consumer this which is expected. Though ideally we should confirm if the memory usage  is expected or extra bloat due to memory not being freed at regular granularity . We should compare with PG for this scenario. 

 CacheMemoryContext: 13098593184 total in 110291 blocks; 11812152 free (48599 chunks); 13086781032 used
 

```
TopMemoryContext: 401416 total in 8 blocks; 5240 free (6 chunks); 396176 used
  Operator class cache: 8192 total in 1 blocks; 552 free (0 chunks); 7640 used
  pgstat TabStatusArray lookup hash table: 8192 total in 1 blocks; 1448 free (0 chunks); 6744 used
  TopTransactionContext: 14204283224 total in 91193 blocks; 7188544 free (452 chunks); 14197094680 used
  TransactionAbortContext: 32768 total in 1 blocks; 32504 free (0 chunks); 264 used
  Portal hash: 8192 total in 1 blocks; 552 free (0 chunks); 7640 used
  TopPortalContext: 8192 total in 1 blocks; 7928 free (0 chunks); 264 used
  Relcache by OID: 8388608 total in 11 blocks; 808616 free (27 chunks); 7579992 used
  CacheMemoryContext: 13098593184 total in 110291 blocks; 11812152 free (48599 chunks); 13086781032 used

.........................
.........................
.........................
.........................

  ErrorContext: 8192 total in 1 blocks; 7928 free (6 chunks); 264 used
Grand total: 31643573792 bytes in 936273 blocks; 220954368 free (53052 chunks); 31422619424 used

```

**Comment by druzac** [comment]
I'm seeing OOMs in the `postgres` backend just on trying to establish a connection to the 70k table database. I was able to perform an initial successful backup after creating the database. I only noticed database connections consistently failing after I tried to restore the backup to the same cluster and that failed.

**Comment by druzac** [comment]
I tried playing around with a similar schema on a pgsql database. Connecting to the database is instantaneous and memory consumption of the `postgres` process is very low, on the order of a few megabytes.  Performing similar operations I did on the yb cluster, I ran `pg_dump` and the resulting sql file to create a copy of the database on the same server. `pg_dump` was failing with:
```
pg_dump: [archiver (db)] query failed: ERROR:  out of shared memory
```
After making these tweaks to the pg config, `pg_dump` completed successfully and I was able to run the resulting script to create a copy of the db:
```
shared_buffers = 4GB
max_locks_per_transaction = 1000
```

The highest memory consumption I saw in my testing was while running `pg_dump`, which hit a high water mark of around 6gb.
